### ISSUE TITLE
How to Disable SSL Evaluation for self signed certificates

### ISSUE THREAD

**Issue by stephen-talari** [p2/OAuth2#145]
My API ssl is using self signed certificate. When i call `authorize()` it throws `"An SSL error has occurred and a secure connection to the server cannot be made"`. 

![screen shot 2016-09-27 at 3 40 33 pm](https://cloud.githubusercontent.com/assets/1725945/18869181/d2e3f370-84c8-11e6-95d9-464a0b9494f2.png)

How can i bypass certificate verification for self signed certificates ?

In cURL, it works with `--insecure`

**Comment by p2** [comment]
It's not well documented but you can use [`OAuth2DebugURLSessionDelegate`](https://p2.github.io/OAuth2/Classes/OAuth2DebugURLSessionDelegate.html) for this:

```swift
self.oauth = OAuth2...(settings: [:])
oauth.sessionDelegate = OAuth2DebugURLSessionDelegate(host: "domain.com")
oauth.authorize() { params, error in
    // ...
}
```

**Comment by stephen-talari** [comment]
Tried with localhost/IP, resulted with same error.

Reason for using self signed is, our app is built on top of local machine (Wifi Controller). Private IP's are not issued SSL's from any CA

**Comment by p2** [comment]
`params` is expected to be nil when authorization doesn't succeed – it would contain authorization parameters.

Can you try to debug, set a breakpoint in [OAuth2DebugURLSessionDelegate#42](https://github.com/p2/OAuth2/blob/master/Sources/Base/OAuth2DebugURLSessionDelegate.swift#L43)? Which version are you running, latest master?

**Comment by stephen-talari** [comment]
I am using `p2.OAuth2 (2.3.0)`, breakpoint is not triggered

**Comment by p2** [comment]
Set a break point at [OAuth2Base.swift#220](https://github.com/p2/OAuth2/blob/swift-2.3/Sources/Base/OAuth2Base.swift#L220) to check whether the session is correctly initiated.

**Comment by stephen-talari** [comment]
breakpoint at `if nil == _session`,  does it mean my session isn't properly initiated.

**Comment by p2** [comment]
When you step through, does it create a new session (line 222) and does it use your `sessionDelegate` on that same line?

**Comment by p2** [comment]
I just tried with the 2.3 version, the breakpoint should be set in [OAuth2Request.swift#71](https://github.com/p2/OAuth2/blob/swift-2.3/Sources/Base/OAuth2Request.swift#L71). It breaks for me, so this should work. I'm not sure if the IP-based thing works, though, maybe it must be a domain name, but it should still break on line 71.

**Comment by stephen-talari** [comment]
[https://github.com/p2/OAuth2/blob/swift-2.3/Sources/Base/OAuth2Base.swift#L224](url)
![screen shot 2016-09-28 at 2 45 58 pm](https://cloud.githubusercontent.com/assets/1725945/18907701/57947556-858a-11e6-846d-92ad5ff5ad8d.png)

**Comment by p2** [comment]
Did you get this to work?

**Comment by stephen-talari** [comment]
Yes, it did work by creating a Manager with DisableEvaluation policy. 

![screen shot 2016-11-21 at 2 31 30 pm](https://cloud.githubusercontent.com/assets/1725945/20476514/4be5d32a-aff7-11e6-8093-576887a81ff2.png)

**Comment by p2** [comment]
Great, thanks for the code example!

**Comment by aliasdoc** [comment]
Hi, I have the same problem, OAuth2DebugURLSessionDelegate() not work, my code:

```

client!.sessionDelegate = OAuth2DebugURLSessionDelegate(host: "xxx.dev")
        client!.authorize() { authParameters, error in
            if let params = authParameters {
                print("Authorized! Access token is in `oauth2.accessToken`")
                print("Authorized! Additional parameters: \(params)")
            }
            else {
                print("Authorization was cancelled or went wrong: \(error)")   // error will not be nil
            }
        }


```

Any idea ? Thanks.

**Comment by p2** [comment]
Are the breakpoints mentioned above being called?

**Comment by aliasdoc** [comment]
Hi @p2 , no it's the same situation, I'm also trying with custom manager but no effect. I thinks the problem is with ATS but I have no solution. I'm using a self-signed certificate generated by MAMP and another generated manually with no effect.

**Comment by p2** [comment]
Which version of the OAuth2 framework are you using, and how (Pod, Carthage, source)?

**Comment by aliasdoc** [comment]
I'm using `OAuthSwift (1.1.0)` (Podfile.lock), my Podfile: `pod 'OAuthSwift', '~> 1.1.0'`, thanks you.

**Comment by p2** [comment]
@aliasdoc Were you able to resolve this issue? If not, do you have a public-facing server with the self-signed certificate so I can test with?

**Comment by jupl** [comment]
I have been attempting to do this with `OAuth2DataLoader` with ATS and `OAuth2DebugURLSessionDelegate` only to be greeted with the error:

```
"The certificate for this server is invalid. You might be connecting to a server that is pretending to be "*******************" which could put your confidential information at risk."
```

I don't have control over the server I'm trying to access. (it's internal work stuff) I'll probably try using Alamofire and see if I can get it to work as someone got success with it above.

**Comment by p2** [comment]
Are you on latest `master` or using CocoaPods? Try latest master as it contains a change to this, which is not yet podded.

**Comment by p2** [comment]
This is now in the latest Pod.

**Comment by aliasdoc** [comment]
Just FYI, to use `OAuth2DebugURLSessionDelegate() ` with self-signed certificate, you need to set your ATS like this:
```xml
<key>NSAppTransportSecurity</key>
<dict>
	<key>NSAllowsArbitraryLoads</key>
	<true/>
</dict>
```

**And only THIS**, if you specify some other keys, sessionDelegate are not fired.

**Comment by shashank109** [comment]
Hi.... I am also facing same issue. can someone pls help me out of this?Here is my code-

                let oauth2 = OAuth2CodeGrant(settings: [
                    "client_id": "codeclient",
                    "client_secret": "secret",
                    "authorize_uri": "https://192.168.0.21:4532/core/connect/authorize",
                    "token_uri": "https://192.168.0.21:4532/core/connect/token",   // code grant only
                    "redirect_uris": ["https://localhost:3000/account/oauth2"],   // register your own "myapp" scheme in Info.plist
                    "scope": "openid profile email role offline_access read write",
                    "secret_in_body": true,    // Github needs this
                    "keychain": false,         // if you DON'T want keychain integration
                    ] as OAuth2JSON)

                let base = URL(string: "https://192.168.0.21:4532/core/.well-known/openid-configuration")!

                oauth2.sessionDelegate = OAuth2DebugURLSessionDelegate(host: "192.168.0.21:4532")
                var req = oauth2.request(forURL: base)
                
                req.setValue("application/vnd.github.v3+json", forHTTPHeaderField: "Accept")
                let oauthloader: OAuth2DataLoader
                oauthloader = OAuth2DataLoader(oauth2: oauth2)
                oauthloader.perform(request: req) { response in
                    do {
                        let dict = try response.responseJSON()
                        DispatchQueue.main.async {
                            // you have received `dict` JSON data!
                        }
                    }
                    catch let error {
                        DispatchQueue.main.async {
                            // an error occurred
                        }
                    }
                }

**Comment by musaurabh** [comment]
Adding a self generated trusted root certificate doesn't work. What's the best way here? To bypass the check or make a ssl cert trusted?
